csrc/multidevice/lower_communication.cpp

```diff
@@ -0,0 +1,324 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#ifdef USE_DISTRIBUTED
+#include <ir/interface_nodes.h>
+#include <multidevice/device_mesh.h>
+#include <multidevice/lower_communication.h>
+#include <multidevice/pipeline.h>
+
+namespace nvfuser {
+
+namespace {
+
+// Returns whether a TensorView has its first axis parallelized on Didx
+// Checks that the other axis are not parallelized on Didx
+bool isParallelD(TensorView* tv) {
+  std::vector<bool> is_parallel_d;
+  for (IterDomain* id : tv->getRootDomain()) {
+    is_parallel_d.push_back(isParallelTypeDeviceDim(id->getParallelType()));
+  }
+  // Currently, only the most external dim is allowed to be parallelized
+  NVF_ERROR(tv->getMaybeRFactorDomain() == tv->getLeafDomain());
+  for (auto i : c10::irange(1, is_parallel_d.size())) {
+    NVF_ERROR(
+        !is_parallel_d.at(i),
+        "only the outmost dimension can be device-parallelized");
+  }
+  return is_parallel_d.empty() ? false : is_parallel_d.at(0);
+}
+
+inline bool isDeviceInvolved(
+    DeviceIdxType device_index,
+    DeviceIdxType root,
+    DeviceMesh mesh) {
+  return device_index == root || mesh.has(device_index);
+}
+
+inline bool isDeviceInvolved(
+    DeviceIdxType device_index,
+    DeviceMesh sender_mesh,
+    DeviceMesh receiver_mesh) {
+  return sender_mesh.has(device_index) || receiver_mesh.has(device_index);
+}
+
+// Creates a dummy tensor for scatter/gather communications,
+// see 'CreateParamsForGatherScatter'
+inline at::Tensor createDummyTensor(at::Tensor reference) {
+  return at::empty_like(reference, reference.options());
+}
+
+// Utility function used for setting up a scatter or gather communication
+// params. Since most  of the steps are somewhat similar/opposite in those
+// cases, we gathered the two implementations into one function. The argument
+// "is_scatter" allows to discriminate between scatter and gather
+CommParams CreateParamsForGatherScatter(
+    DeviceIdxType device_index,
+    DeviceIdxType root,
+    DeviceMesh mesh, // is_scatter? receivers : senders
+    at::Tensor root_buf, // is_scatter? input buf : output buf
+    at::Tensor buf, // is_scatter? output buf : input buf
+    bool is_scatter) {
+  CommParams params;
+  params.root = root;
+  params.team = mesh.vector();
+  bool is_root_in_mesh = mesh.has(root);
+  if (!is_root_in_mesh) {
+    params.team.push_back(root);
+  }
```

Comment:
We actually check that there is no duplicate in the team, inside the constructor of `Communication ` https://github.com/samnordmann/Fuser/blob/33501fd04792e69a252262abe1ed60f4dd6e7773/csrc/multidevice/communication.cpp#L63****

Thanks for the suggestion. But I'm afraid this implementation still does not prevent duplicates because of the method `CommParams::setDevices`, right ?
Anyway, I would prefer to leave the DeviceMesh design as light as possible for now in this PR, because I am working on revising it. Notably I think we should make it a symbolic Val. I can add an entry in the TODO list